tqdb/dbr.py

```diff
@@ -40,44 +44,50 @@ def read(dbr):
 
     """
     try:
-        dbr_file = open(dbr)
-    except FileNotFoundError:
-        logging.debug(f'No file found for {dbr}')
-        return {}
-    except PermissionError:
-        logging.debug(f'Tried opening a directory {dbr}')
-        return {}
+        with open(dbr) as dbr_file:
+            result = {}
 
-    result = {}
+            # DBR lines always end with ',\n' which we remove
+            lines = (line.rstrip(',\n') for line in dbr_file)
 
-    # Put all the DBR lines (that always end with ,\n) in a list
-    lines = [line.rstrip(',\n') for line in dbr_file]
+            # Only add properties that have the correct format per line of: key,value
+            properties = dict(tuple(line.split(',', 1)) for line in lines if ',' in line)
 
-    # Only add properties that have the correct format per line of: key,value
-    properties = dict(line.split(',') for line in lines if ',' in line)
+            # The 'templateName' property isn't in any Template, so add manually:
+            if 'templateName' in properties:
+                result['templateName'] = properties['templateName']
 
-    # The 'templateName' property isn't in any Template, so add manually:
-    if 'templateName' in properties:
-        result['templateName'] = properties['templateName']
+            # Get the template for this DBR
+            template = get_template(properties, dbr)
 
-    # Get the template for this DBR
-    template = get_template(properties, dbr)
+            def parse_vars_to_tuples(variables_dict):
+                for name, variable in variables_dict.items():
+                    if name in properties:
+                        parsed_value = variable.parse_value(properties[name])
+                        # Omit None, False, zero, and empty collections
+                        if parsed_value:
+                            yield (name, parsed_value)
 
-    # Now parse all properties using the Template:
-    result.update(dict(
-        (name, variable.parse_value(properties[name])) for
-        name, variable in template.variables.items()
-        if name in properties and
-        variable.parse_value(properties[name])))
+            # Now parse all properties using the Template:
+            result.update(dict(parse_vars_to_tuples(template.variables)))
 
-    return result
+            return result
+
+    except FileNotFoundError:
+        logging.debug(f'No file found for {dbr}')
+        return {}
+    except PermissionError:
+        logging.debug(f'Tried opening a directory {dbr}')
+        return {}
 
 
-def parse(dbr_file, references={}):
+def parse(dbr_file, references=None):
     """
     Parse a DBR file according to its template.
 
     """
+    if references is None:
```

Comment:
Did I do default values wrong? 😅

Reply:
This change was an automatic refactor suggested by IntelliJ IDEA. It claimed that using a mutable default parameter was dangerous because if that value ever got modified, it would impact future calls. I admit I am not sure if it was a real problem because that would require trying to determine what happens to the return value. But this does seem like it's safe.

Reply:
Cool, TIL 👍